File: C++/102_Binary Tree Level Order Traversal.cpp


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root==nullptr){
            return {};
        }
        deque<TreeNode*> q;
        vector<vector<int>> result;
        int level=0;
        int l;
        TreeNode* node = root;
        q.push_back(node);
        while (!q.empty()){
            l = q.size();
            result.push_back({});
            for (int i=0;i<l;i++){
                node = q.front();
                q.pop_front();
                result[level].push_back(node->val);
                if (node->left != nullptr){
                    q.push_back(node->left);
                }
                if (node->right != nullptr){
                    q.push_back(node->right);
                }
            }
            level += 1;

        }

        return result;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (root == nullptr){
            return {};
        }
        queue<TreeNode*> q;
        vector<vector<int>> result;
        TreeNode* node = root;
        q.push(node);
        int level = 0;
        int l;
        while (!q.empty()){
            l = q.size();
            result.push_back({});
            for (int i=0;i<l;i++){
                node = q.front();
                q.pop();
                result[level].push_back(node->val);
                if (node->left != nullptr){
                    q.push(node->left);
                }
                if (node->right != nullptr){
                    q.push(node->right);
                }

            }
            level += 1;

        }
        return result;

    }
};

File: Python/102_Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        q = collections.deque([root])
        res = []

        while q:
            cur_level, n = [], len(q)
            for i in range(n):
                node = q.popleft()
                if node:
                    cur_level.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(cur_level)
        return res

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
            
        q = collections.deque([root])
        res = []

        while q:
            cur_level, n = [], len(q)
            for i in range(n):
                node = q.popleft()
                if node:
                    cur_level.append(node.val)
                    if node.left:
                        q.append(node.left)
                    if node.right:
                        q.append(node.right)
            res.append(cur_level)
        return res[::-1]